File: C++/Medium/RuntimeError/maximum-non-negative-product-in-a-matrix_1.cpp
class Solution {
public:
    int maxProductPath(vector<vector<int>>& grid) {
        
        int m = grid.size();
        int n = grid[0].size();
        vector<vector<int>> maxProd(m, vector<int>(n, 0));
        vector<vector<int>> minProd(m, vector<int>(n, 0));
        maxProd[0][0] = minProd[0][0] = grid[0][0];
        for (int i = 1; i < m; ++i) {
            maxProd[i][0] = maxProd[i - 1][0] * grid[i][0];
            minProd[i][0] = minProd[i - 1][0] * grid[i][0];
        }
        for (int j = 1; j < n; ++j) {
            maxProd[0][j] = maxProd[0][j - 1] * grid[0][j];
            minProd[0][j] = minProd[0][j - 1] * grid[0][j];
        }
        for (int i = 1; i < m; ++i) {
            for (int j = 1; j < n; ++j) {
                if (grid[i][j] > 0) {
                    maxProd[i][j] = max(maxProd[i - 1][j], maxProd[i][j - 1]) * grid[i][j];
                    minProd[i][j] = min(minProd[i - 1][j], minProd[i][j - 1]) * grid[i][j];
                } else {
                    maxProd[i][j] = min(minProd[i - 1][j], minProd[i][j - 1]) * grid[i][j];
                    minProd[i][j] = max(maxProd[i - 1][j], maxProd[i][j - 1]) * grid[i][j];
                }
            }
        }
        return maxProd[m - 1][n - 1] < 0 ? -1 : maxProd[m - 1][n - 1] % 1000000007;
    }
};

File: Java/Medium/Accepted/maximum-non-negative-product-in-a-matrix_1.java
class Solution {
    public int maxProductPath(int[][] grid) {
        
        int m = grid.length;
        int n = grid[0].length;
        long[][] max = new long[m][n];
        long[][] min = new long[m][n];
        max[0][0] = grid[0][0];
        min[0][0] = grid[0][0];
        for (int i = 1; i < m; i++) {
            max[i][0] = max[i - 1][0] * grid[i][0];
            min[i][0] = min[i - 1][0] * grid[i][0];
        }
        for (int j = 1; j < n; j++) {
            max[0][j] = max[0][j - 1] * grid[0][j];
            min[0][j] = min[0][j - 1] * grid[0][j];
        }
        for (int i = 1; i < m; i++) {
            for (int j = 1; j < n; j++) {
                long[] tmp = new long[4];
                tmp[0] = max[i - 1][j] * grid[i][j];
                tmp[1] = max[i][j - 1] * grid[i][j];
                tmp[2] = min[i - 1][j] * grid[i][j];
                tmp[3] = min[i][j - 1] * grid[i][j];
                max[i][j] = Arrays.stream(tmp).max().getAsLong();
                min[i][j] = Arrays.stream(tmp).min().getAsLong();
            }
        }
        return max[m - 1][n - 1] < 0 ? -1 : (int) (max[m - 1][n - 1] % 1000000007);
    }
}